### PR TITLE
feat: 🎸 add new 3.1.0 methods for GPM

### PR DESCRIPTION
**Pull request by shuffledex**
add addDelegateMulti and removeDelegateMulti presents on 3.1.0 contracts

### CODE REVIEW

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.0.0.ts [hunk 1]
```diff
@@ -152,6 +155,22 @@ interface ChangePermissionMultiParams extends TxParams {
   valids: boolean[];
 }
 
+/**
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.0.0.ts]: `AddDelegateMultiParams` and `DeleteDelegateMultiParams` shouldn't be here. They are from 3.1.0 version.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/__tests__/3.1.0.test.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  });
+
+  afterEach(() => {
+    reset(mockedWrapper);
+    reset(mockedContract);
+    reset(mockedSecurityTokenContract);
+    reset(mockedContractFactory);
+  });
+
+  describe('addDelegateMulti', () => {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/__tests__/3.1.0.test.ts]: The constructor test is missed. It should check if GPM is extending GPM_Common

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/index.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+export type GeneralPermissionManager = GeneralPermissionManager_3_0_0 | GeneralPermissionManager_3_1_0;
+
+export {
+  GeneralPermissionManager_3_0_0,
+  isGeneralPermissionManager_3_0_0,
+  GeneralPermissionManager_3_1_0,
+  isGeneralPermissionManager_3_1_0,
+};
+export { default as GeneralPermissionManagerCommon } from './common';
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/index.ts]: It should export `GeneralPermissionManagerTransactionParams` from `./common` also.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/common.ts [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+  modules: string[];
+  perms: Perm[];
+  valids: boolean[];
+}
+
+/**
+ * @param delegates An array of Ethereum addresses of the delegates
+ * @param details An array of details about the delegates i.e `Belongs to financial firm`
+ */
+interface AddDelegateMultiParams extends TxParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/common.ts]: This is not needed in common as it is implemented only in 3.1.0

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/common.ts [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  delegates: string[];
+  details: string[];
+}
+
+/**
+ * @param delegates An array of Ethereum address of delegates
+ */
+interface DeleteDelegateMultiParams extends TxParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/common.ts]: This is not needed in common as it is implemented only in 3.1.0

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+interface GetGeneralPermissionManagerLogsAsyncParams extends GetLogs {
+  (params: GetChangePermissionLogsAsyncParams): Promise<
+    LogWithDecodedArgs<GeneralPermissionManagerChangePermissionEventArgs_3_1_0>[]
+  >;
+  (params: GetAddDelegateLogsAsyncParams): Promise<
+    LogWithDecodedArgs<GeneralPermissionManagerAddDelegateEventArgs_3_1_0>[]
+  >;
+}
+
+export namespace GeneralPermissionManagerTransactionParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts]: This is already exported in `common`, so we don't need to export it again.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+  export interface DeleteDelegate extends DelegateTxParams {}
+  export interface AddDelegate extends AddDelegateParams {}
+  export interface ChangePermission extends ChangePermissionParams {}
+  export interface ChangePermissionMulti extends ChangePermissionMultiParams {}
+}
+
+/**
+ * @param delegate Ethereum address of the delegate
+ */
+interface DelegateTxParams extends TxParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts]: Once the previous export is deleted, this is not necessary any more.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+  delegate: string;
+}
+
+/**
+ * @param delegate Ethereum address of the delegate
+ * @param details Details about the delegate i.e `Belongs to financial firm`
+ */
+interface AddDelegateParams extends TxParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts]: Once the previous export is deleted, this is not necessary any more.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+  details: string;
+}
+
+/**
+ * @param delegate Ethereum address of the delegate
+ * @param module Ethereum contract address of the module
+ * @param perm Permission flag
+ * @param valid Bool flag use to switch on/off the permission
+ */
+interface ChangePermissionParams extends TxParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts]: Once the previous export is deleted, this is not necessary any more.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+  valid: boolean;
+}
+
+/**
+ * @param delegate Ethereum address of the delegate
+ * @param modules Multiple module matching the multiple perms, needs to be same length
+ * @param perms Multiple permission flag needs to be changed
+ * @param valids Bool array consist the flag to switch on/off the permission
+ */
+interface ChangePermissionMultiParams extends TxParams {
```
Comment (VictorVicente) [inline src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/3.1.0.ts]: Once the previous export is deleted, this is not necessary any more.

File: src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/index.ts (hunk 3 above)
Comment (shuffledex) [reply src/contract_wrappers/modules/permission_manager/general_permission_manager_wrapper/index.ts]: There is other task regarding this. I proceed to close this issue.

**Pull request merged**